lib/harvesting/projects.exs

```diff
@@ -10,12 +10,14 @@ defmodule CLI do
       {:ok, _datetime} = result ->
         result
       {:error, :invalid_format} ->
-        case Date.from_iso8601(date_string) do
-          {:ok, _date} ->
-            DateTime.from_iso8601("#{date_string}T00:00:00Z")
-          error ->
-            error
-        end
+        parse_arguments("#{date_string}T00:00:00Z")
+    end
+  end
+
+  def parse_arguments(date_string) do
```

Comment:
Das hier heißt dann aber auch, dass man den Parameter `--script` nicht mehr braucht, um das Skript zu starten, oder? Ist das beabsichtigt?

Reply:
Nein den Skript Parameter brauchst du weiterhin. Die Consolen-Argumente werden als Liste übergeben, wenn du es ohne --skript aufrufst, übergibst du also [date_string] und dann greift dieses Matching nicht.

Reply:
Dazu: https://github.com/dainst/argos/blob/main/lib/application.ex#L22

Reply:
Also das --script ist nur ein Flag, so dass mix die Application children (Cowboy + Harvester) nicht startet.

Reply:
(Es mag eine elegantere Variante geben)

Reply:
Ah, jetzt verstehe ich worüber David gestolpert ist:

Die Funktion sollte erst einmal weiter nach oben unter die anderen parse_arguments (an sich sollte es da auch ein Compiler Warning geben). 

Was etwas verwirrend ist: Es gibt jetzt 2x `parse_arguments`, die eine Liste mit zwei Elementen als Parameter erwarten und 1x `parse_arguments`, wo nur ein string erwartet wird. Der pattern-matched in Nils code entsprechend durch und landet bei letzterem. Das "--script" wurde zu dem Zeitpunkt bereits in einer der beiden vorhergehenden Funktionen gematched und nicht weiter gereicht.

Reply:
Wobei der Name `parse_arguments` für die neue Funktion nicht so richtig passt: Zu dem Zeitpunkt parsen wir nix mehr was direkt aus dem Konsolenaufruf kommt, sondern nur einen neu konstruierten String. Vielleicht könnte man die Logik einfach mit in `parse_arguments(["--script", date_string])` integrieren?

Reply:
> Das hier heißt dann aber auch, dass man den Parameter `--script` nicht mehr braucht, um das Skript zu starten, oder? Ist das beabsichtigt?

Antwort: Nein, da von der Konsole eine Liste kommt, die neue Funktion da also nie matched.

Reply:
> Mist, mir fällt gerade etwas auf, was auch beim Chronontology ein Problem sei könnte. Es gibt ein Limit bei der Ergebnisgröße in Elasticsearch, was den Offset angeht.

Das ist kein Problem. Dar Harvester macht bereits ein eigenes Paging hier:

https://github.com/dainst/argos/blob/main/lib/harvesting/chronontology.ex#L71

Reply:
Mmh, ok, also wenn man den Code unbefangen liest, sieht es so aus, als könnte man da auch direkt einen date_string übergeben als Parameter für das Skript. Ich habe das gerade mal ausprobiert und er springt auch tatsächlich in die Funktion, wenn man einen Aufruf ohne --skript aber mit ISO-Date macht.

Ich finde die Funktion sollte anders heißen, weil sie auf einer anderen Ebene ist als die beiden "parse_argument()", die direkt UserInput behandeln. Warum nicht `parse_date`?

Reply:
> Ah, jetzt verstehe ich worüber David gestolpert ist:
> 
> Die Funktion sollte erst einmal weiter nach oben unter die anderen parse_arguments (an sich sollte es da auch ein Compiler Warning geben).
> 
> Was etwas verwirrend ist: Es gibt jetzt 2x `parse_arguments`, die eine Liste als Parameter erwarten und 1x `parse_arguments`, wo nur ein string erwartet wird. Der pattern-matched in Nils code entsprechend durch und landet bei letzterem. Das "--script" wurde zu dem Zeitpunkt bereits in einer der beiden vorhergehenden Funktionen gematched und nicht weiter gereicht.

Die Funktion steht bei den anderen parse_arguments?! Ich vermute es ist ein Missverständnis.

Reply:
> Wobei der Name `parse_arguments` für die neue Funktion nicht so richtig passt: Zu dem Zeitpunkt parsen wir nix mehr was direkt aus dem Konsolenaufruf kommt, sondern nur einen neu konstruierten String. Vielleicht könnte man die Logik einfach mit in `parse_arguments(["--script", date_string])` integrieren?

Es war vorher in der oberen Funktion integriert, das war aber sehr verschachtelt und ich wollte ein Lösung mit Pattern-Match finden.

Reply:
> Wobei der Name `parse_arguments` für die neue Funktion nicht so richtig passt: Zu dem Zeitpunkt parsen wir nix mehr was direkt aus dem Konsolenaufruf kommt, sondern nur einen neu konstruierten String. Vielleicht könnte man die Logik einfach mit in `parse_arguments(["--script", date_string])` integrieren?

Man könnte der Funktion einen eigenen Namen geben und das ganze als private Funktion lösen, das wäre dann aber wieder sehr prozeduraler Code. 

Da der Input zwischendrin eigentlich nicht mehr manipuliert werden kann, da diese Funktion nur durch die Rekursion aufgerufen wird, befindet man sich noch im Prozess des Argumentparsings, deshalb habe ich den Funktionsnamen beibehalten.

Reply:
Das Überladen des Funktionsnamens ist halt das Grundprinzip von diesem Patternmatching, deshalb wollte ich das hier beibehalten. 

> Mmh, ok, also wenn man den Code unbefangen liest, sieht es so aus, als könnte man da auch direkt einen date_string übergeben als Parameter für das Skript. Ich habe das gerade mal ausprobiert und er springt auch tatsächlich in die Funktion, wenn man einen Aufruf ohne --skript aber mit ISO-Date macht.
> 
> Ich finde die Funktion sollte anders heißen, weil sie auf einer anderen Ebene ist als die beiden "parse_argument()", die direkt UserInput behandeln. Warum nicht `parse_date`?

Kann ich nicht reproduzieren:
` mix run lib/harvesting/projects.exs 2020-01-10T00:00:00Z` oder
` mix run lib/harvesting/projects.exs 2020-01-10`

führen beide zu ** (FunctionClauseError) no function clause matching, wie es zu erwarten war.

Reply:
Ja, ist ein FunctionClauseError, aber in einer späteren Funktion, nämlich parse_utc_datetime. Er springt erst in die parse_arguments und das ist ja eigentlich nicht beabsichtigt, oder? Ich glaube insgesamt, die Datei wäre besser zu verstehen, wenn man die Funktion anders benennt. Ich kann aber auch damit leben, wenn es so bleibt.

Reply:
Wie weiter unten bereits geschrieben, habe ich diese Änderung wieder zum vorherigen Stand reverted.

Reply:
> Ja, ist ein FunctionClauseError, aber in einer späteren Funktion, nämlich parse_utc_datetime. Er springt erst in die parse_arguments und das ist ja eigentlich nicht beabsichtigt, oder? Ich glaube insgesamt, die Datei wäre besser zu verstehen, wenn man die Funktion anders benennt. Ich kann aber auch damit leben, wenn es so bleibt.

Ja, du hast recht. Der von mir gedachte patter_match hat hier nicht funktioniert, weil date_string selbst keinen Type verlangte und damit einfach alles akzeptiert hat auch die Liste. Man hätte das vermutlich mit einer Guard verhindern können.